### PR TITLE
Protorev: Update timing for pool updating and dev fee payout

### PR DESCRIPTION
**Pull request by NotJeremyLiu**
## What is the purpose of the change

This PR updates the timing cadence of certain protorev logic:
1. Changes dev account payment from once per week to after every trade (this is because the accumulation into the module account and taking from it has bad UX for those tracking protorev profits)
2. Changes highest liquidity pools updating from once per week to daily
3. Sets up protorev module for v16 upgrade migration

## Brief Changelog

- Uses daily epoch instead of weekly epoch for highest liquidity updating
- Removes all developer fee logic from epoch triggering into ExecuteTrade triggering
- Adds an upgrade migration in ProtoRev to payout what is left in the store when the v16 upgrade happens. This store will no longer be used with the new logic
- Marks (via comments) no longer used functions in v16 as deprecated , and also denotes functions used only for the upgrade as removable in v17

## Testing and Verifying

- Removed tests for no longer used methods
- Added a test for the new SendDeveloperFee method
- All other tests pass after setting the developer account in the tests
- Added checks in the TestUpgrade() function in v16 upgrade_test.go to ensure protorev upgrades properly (transfers all leftover dev fee funds in the store the the dev account)

## Documentation and Release Note
- Changelog added
- Documentation has been updated to reflect new changes

### CODE REVIEW

**Comment by NotJeremyLiu** [comment]
This commit is a minimal code change version of sending developer profit after every trade: 5bacd9ecbace4f968367a5d4589bc5ffb02bb3a2

It's not optimized for this change since it still stores state to track the dev fee (which is now unnecessary). Will be changed in a future commit to be an optimal version

**Comment by bpiv400** [comment]
Should we interpret this as unfinished until the unnecessary state tracking is removed?

**Review by davidterpay — APPROVED** [review]
Not too familiar with what migration of developer fees away from ProtoRev in v16 looks like, so I would double check with the team. However, core changes LGTM.

**Comment by NotJeremyLiu** [comment]
R4R besides merge conflicts

**Comment by NotJeremyLiu** [comment]
Note to reviewers: The part I'm most unsure about is the upgrade/migration logic. Please take a careful look at that, thank you! I got it to work in testing through trial and error

**Comment by NotJeremyLiu** [comment]
merge conflicts resolved

**Comment by NotJeremyLiu** [comment]
Another note: The e2e tests are failing, I imagine related to the way I handled upgrades? Will need some assistance from Osmosis team on this.

**Comment by stackman27** [comment]
> Another note: The e2e tests are failing, I imagine related to the way I handled upgrades? Will need some assistance from Osmosis team on this.

yup looking into this rn

File: app/upgrades/v16/upgrades.go [hunk 1]
```diff
@@ -128,6 +128,12 @@ func CreateUpgradeHandler(
 
 		updateTokenFactoryParams(ctx, keepers.TokenFactoryKeeper)
 
+		// Transfers out all the dev fees in kvstore to dev account during upgrade
+		// Expceted to error during e2e test due to dev account not being set
```
Comment (p0mvn) [inline app/upgrades/v16/upgrades.go:132]:
```suggestion
		// Expected to error during e2e test due to dev account not being set
```

File: app/upgrades/v16/upgrades.go [hunk 2]
```diff
@@ -128,6 +128,12 @@ func CreateUpgradeHandler(
...
+		if err := keepers.ProtoRevKeeper.SendDeveloperFeesToDeveloperAccount(ctx); err != nil {
+			ctx.Logger().Error("error sending developer fees to developer account during upgrade, expected in testing, not in prod", "error", err)
+		}
```
Comment (p0mvn) [inline app/upgrades/v16/upgrades.go:135]: I would prefer it if this is handled properly. I know the process for updating e2e is cumbersome. However, handling this properly will save us time during the upgrade if any issues occur.

The process for setting this account in e2e is the following:

1. update e2e genesis for the protorev module in a separate PR: https://github.com/osmosis-labs/osmosis/blob/1e56cc620a17e0b5debb8228896aa44b6421a723/tests/e2e/initialization/config.go#L234-L237
2. backport that PR to `v15.x`
3. Grab the chain init docker image with v15 tag from here: https://hub.docker.com/r/osmolabs/osmosis-e2e-init-chain/tags
  * should be auto-uploaded
4. Copy this tag: https://github.com/osmosis-labs/osmosis/blob/1e56cc620a17e0b5debb8228896aa44b6421a723/tests/e2e/containers/config.go#L27

Comment (stackman27) [reply app/upgrades/v16/upgrades.go:135]: hi @NotJeremyLiu is this PR r4r now. I think just needs rebasing right?

Comment (NotJeremyLiu) [reply app/upgrades/v16/upgrades.go:135]: @stackman27 I just updated the tag and removed the failing test, letting it run through workflow to see if all is good

File: tests/e2e/containers/config.go [hunk 3]
```diff
@@ -24,7 +24,7 @@ const (
 	// It should be uploaded to Docker Hub. OSMOSIS_E2E_SKIP_UPGRADE should be unset
 	// for this functionality to be used.
 	previousVersionOsmoRepository = "osmolabs/osmosis-dev"
-	previousVersionOsmoTag        = "v15.x-833705e4-1679340262"
+	previousVersionOsmoTag        = "v15.x-63e3f053-1685053483"
```
Comment (p0mvn) [inline tests/e2e/containers/config.go:27]: please also bump `previousVersionInitTag` below to v15.x-833705e4-1679340262

**Comment by NotJeremyLiu** [comment]
All checks pass, finally r4r @stackman27

**Review by stackman27 — APPROVED** [review]
LGTM

File: x/protorev/keeper/developer_fees.go [hunk 4]
```diff
@@ -32,6 +33,7 @@ func (k Keeper) SendDeveloperFeesToDeveloperAccount(ctx sdk.Context) error {
 	return nil
 }
 
+// Deprecated: Can be removed in v16
 // UpdateDeveloperFees updates the fees that developers can withdraw from the module account
 func (k Keeper) UpdateDeveloperFees(ctx sdk.Context, denom string, profit sdk.Int) error {
```
Comment (p0mvn) [inline x/protorev/keeper/developer_fees.go:38]: Let's remove now?

**Review by p0mvn — APPROVED** [review]
LGTM

File: x/protorev/keeper/developer_fees.go (hunk 4 above)
Comment (p0mvn) [reply x/protorev/keeper/developer_fees.go:38]: This is unused now. Please submit separate PR if agreed. Not going to block merge on this

**Pull request merged**